### PR TITLE
Improved performance of product category rewrites (disabled product & categories and product_use_categories)

### PR DESCRIPTION
**Pull request by fballiano**
This PR is a new version of https://github.com/OpenMage/magento-lts/pull/1405 since the original one was not editable by admin, was on the wrong branch and there was no feedback since a while.

@colinmollenhour I've also applied your PR https://github.com/macopedia/magento-lts/pull/8 here in this code already.

### CODE REVIEW

File: app/code/core/Mage/Catalog/Model/Resource/Url.php
Comment (fballiano) [inline app/code/core/Mage/Catalog/Model/Resource/Url.php:1036]: @colinmollenhour since we're modifying a public method, we should base it on "next" and consider it a BC no?

Comment (colinmollenhour) [reply app/code/core/Mage/Catalog/Model/Resource/Url.php:1036]: In my opinion adding an argument with a default value is not a big BC risk since users calling that method in their own code properly should not be affected. Now there could be third-party extensions that already overrode it but it seems very unlikely. So I think "main" is ok in this case.

**Comment by fballiano** [comment]
I'll manually add the you and him in the commit message if that's ok

**Comment by colinmollenhour** [comment]
Are you talking about adding "@" with teh github username to the commit message? It's not a big deal but I think that is not the best.. I end up getting emails over time when those commits are rebased and pushed even on other projects - also I think it might not update the Contributors section in the README automatically to add Marcin. But again, not blocker per-se.

**Comment by fballiano** [comment]
no I add the "coauthored by" message in the squash commit, I always do

**Review by empiricompany — DISMISSED** [review]
I did some testing and looks good to me

**Comment by empiricompany** [comment]
> I did some testing and looks good to me

The only thing is that it doesn't prevent generating rewrites for child categories of a disabled category

**Comment by empiricompany** [comment]
this and mine https://github.com/OpenMage/magento-lts/pull/3251
should be merged into next rc for me, it's an important feature of perfomance improvement

**Comment by fballiano** [comment]
this PR has too many problems that have to be addressed before merging, it's also a quite big one so I think it should go to a later release.

@MarcinNowakMacopedia are you still interested in checking what's reported by phpstan?

**Comment by ADDISON74** [comment]
Most likely such PRs and those marked with "Don't Forget" will have to be promoted by a maintainer so that they can be corrected and approved by the others.

**Comment by fballiano** [comment]
Because of https://github.com/OpenMage/magento-lts/pull/3267#issuecomment-1619976476 I'm closing this PR.

**Comment by ysrtech** [comment]
This PR makes a huge improvement for indexing preformance. Please lets get this fixed and merged.

File: app/code/core/Mage/Catalog/Model/Url.php
Comment (ysrtech) [reply app/code/core/Mage/Catalog/Model/Url.php:988]: Yes this can be removed.

File: app/code/core/Mage/Catalog/Model/Url.php
Comment (ysrtech) [reply app/code/core/Mage/Catalog/Model/Url.php:367]: Its safe to revert it back to the original while ($process == true)

**Pull request closed without merging**
